packages/flutter/lib/src/widgets/lazy_block.dart

```diff
@@ -0,0 +1,449 @@
+
+  @override
+  _RenderLazyBlock createRenderObject(BuildContext context) => new _RenderLazyBlock();
+}
+
+class _LazyBlockParentData extends ContainerBoxParentDataMixin<RenderBox> { }
+
+class _RenderLazyBlock extends RenderVirtualViewport<_LazyBlockParentData> {
+  _RenderLazyBlock({
+    Offset paintOffset: Offset.zero,
+    Axis mainAxis: Axis.vertical,
+    LayoutCallback callback
+  }) : super(
+    paintOffset: paintOffset,
+    mainAxis: mainAxis,
+    callback: callback
+  );
+
+  @override
+  void setupParentData(RenderBox child) {
+    if (child.parentData is! _LazyBlockParentData)
+      child.parentData = new _LazyBlockParentData();
+  }
+
+  @override
+  double getMinIntrinsicWidth(BoxConstraints constraints) {
+    assert(constraints.debugAssertIsNormalized);
+    return constraints.constrainWidth(0.0);
+  }
+
+  @override
+  double getMaxIntrinsicWidth(BoxConstraints constraints) {
+    assert(constraints.debugAssertIsNormalized);
+    return constraints.constrainWidth(0.0);
+  }
+
+  @override
+  double getMinIntrinsicHeight(BoxConstraints constraints) {
+    assert(constraints.debugAssertIsNormalized);
+    return constraints.constrainHeight(0.0);
+  }
+
+  @override
+  double getMaxIntrinsicHeight(BoxConstraints constraints) {
+    assert(constraints.debugAssertIsNormalized);
+    return constraints.constrainHeight(0.0);
+  }
+
+  @override
+  bool get sizedByParent => true;
+
+  @override
+  bool get isRepaintBoundary => true;
+
+  @override
+  void performResize() {
+    size = constraints.biggest;
+  }
+
+  @override
+  void performLayout() {
+    if (callback != null)
+      invokeLayoutCallback(callback);
+  }
+}
+
+class _LazyBlockElement extends RenderObjectElement {
+  _LazyBlockElement(LazyBlockViewport widget) : super(widget);
+
+  @override
+  LazyBlockViewport get widget => super.widget;
+
+  @override
+  _RenderLazyBlock get renderObject => super.renderObject;
+
+  double _firstChildLogicalOffset = 0.0;
+  int _firstChildLogicalIndex = 0;
+  List<Element> _children = <Element>[];
+
+  double _minScrollOffset = 0.0;
+  double _maxScrollOffset = 0.0;
+
+  double _startOffsetLowerLimit = 0.0;
+  double _startOffsetUpperLimit = 0.0;
+
+  @override
+  void visitChildren(ElementVisitor visitor) {
+    for (Element child in _children)
+      visitor(child);
+  }
+
+  @override
+  void mount(Element parent, dynamic newSlot) {
+    super.mount(parent, newSlot);
+    renderObject
+      ..callback = _layout
+      ..mainAxis = widget.direction;
+    // Children will get built during layout.
+    // Paint offset will get updated during layout.
+  }
+
+  @override
+  void update(LazyBlockViewport newWidget) {
+    LazyBlockViewport oldWidget = widget;
+    super.update(newWidget);
+    renderObject.mainAxis = widget.direction;
+    if (newWidget.delegate.runtimeType != oldWidget.delegate.runtimeType
+        || newWidget.delegate.shouldRebuild(oldWidget.delegate)) {
```

Comment:
the `||` should be on the previous line if having it on the next line doesn't line things up nicely